### ISSUE TITLE
[Question] Ability to draw only updated area of a View (partial/incremental updates)

### ISSUE THREAD

**Issue by wu000227** [gyscos/Cursive#349]
I am trying to draw some complex things on a custom ```View```, everything is working; however, I find it is usually a situation that I want to change few random things on the ```View``` while the majority of the texts are still remaining the same.

My question is, if there are such options that allow you to update only certain areas of a ```View```? If no, would it be possible to be supported in the future? It can be a huge performance improvement for running apps that output complex things on a terminal app if it is supported.

Another question is, is there any way to call the ```draw``` method of a ```View``` manually? My approach is to receive data from a ```channel```, and then update the ```View``` regarding to what were received. My current solution is to enable the ```auto_refresh``` trigger so that I can handle & receive data inside ```on_event```, but it doesn't feel like the right way for me. What would be the recommend way to do this using Cursive?

**Comment by gyscos** [comment]
Thanks for the report!

> My question is, if there are such options that allow you to update only certain areas of a View?

Currently there's no global support for partial/incremental updates. It means each view must re-draw its entire content in the `View::draw` step. Adding partial refreshes would indeed save some data sent to the terminal, but most of the time bandwidth isn't such a big issue. It certainly would be a good feature to have, but will need some design work to come up with an API both efficient and convenient to use. So it may happen some day, but I don't have a timeline for that yet.

What can be costly is if your view needs complex computation to even _know_ what to draw - for example, a simple `TextView` may need to wrap around text and compute line boundaries, a `LinearLayout` may need to compute the size for each of its children, ...
The usual solution to this is to cache some intermediate step. For example, cache the state of each component and only re-compute the state of components that changed.

> Another question is, is there any way to call the draw method of a View manually?

You can trigger a screen refresh asynchronously: `Cursive::cb_sink()` will give you a channel that you can use to send callbacks. Anytime a callback is executed, the screen is also refreshed; so you can send a dummy callback like `Box::new(Cursive::noop)` to trigger a refresh. You can also do this from another thread!

**Comment by wu000227** [comment]
Thanks for the reply, this answers my question.


Here is a note for people who might be facing screen blinking/terminal app high cpu usage due to frequent refreshing the whole screen:

After I changed the backend from the default (ncurses) to **pancurses**, the output became way smoother, and no blinking at all. CPU usage of terminal apps I've tried also dropped from 40~60 percent to less then 10 percent. What's strange is, the CPU usage of the actual program that outputs to terminal also dropped from 16~20 percent to less then 8 percent, while no noticable RAM usage difference. I guess pancurses might somehow buffered the output, so if your program may need a partial/incremental update, pancurses backend can be an option.

Terminal apps I used: iTerm2 & macOS builtin
